rasa/core/brokers/kafka.py

```diff
@@ -209,16 +218,65 @@ def _publish(self, event: Dict[Text, Any]) -> None:
             f" key={partition_key!s}, headers={headers})"
         )
 
+        serialized_event = json.dumps(event).encode(DEFAULT_ENCODING)
+
         if self.producer is not None:
-            self.producer.send(
-                self.topic, value=event, key=partition_key, headers=headers
+            self.producer.poll(0.0)
+            self.producer.produce(
+                self.topic,
+                value=serialized_event,
+                key=partition_key,
+                headers=headers,
+                on_delivery=delivery_report,
             )
 
     def _close(self) -> None:
-        if self.producer is not None:
-            self.producer.close()
+        self._cancelled = True
+        self._poll_thread.join()
 
     @rasa.shared.utils.common.lazy_property
     def rasa_environment(self) -> Optional[Text]:
         """Get value of the `RASA_ENVIRONMENT` environment variable."""
         return os.environ.get("RASA_ENVIRONMENT", "RASA_ENVIRONMENT_NOT_SET")
+
+    def _poll_loop(self) -> None:
+        if self.producer is not None:
+            while not self._cancelled:
+                self.producer.poll(0.1)
```

Comment:
let's add a comment linking to the example this came from + and a sentence or two why we do this (to trigger the callbacks passed to `produce`)

Reply:
Oh in this case, is [this line](https://github.com/RasaHQ/rasa/pull/11765/files#diff-f7a20b1ffcae5c0bc2cfc41728bdb3e107c953434a3e2af23f74f227519d30b5R224) redundant in `_publish` method?

Reply:
Addressed, and removed the polling in `_publish()`: https://github.com/RasaHQ/rasa/pull/11765/commits/c34bc473b75fa51cffed3e1222c224633665f89e